src/charm.py

```diff
@@ -302,20 +369,212 @@ def _on_admin_ingress_ready(self, event: IngressPerAppReadyEvent) -> None:
             logger.info("This app's admin ingress URL: %s", event.url)
 
         self._update_hydra_endpoints_relation_data(event)
+        self._update_endpoint_info()
 
     def _on_public_ingress_ready(self, event: IngressPerAppReadyEvent) -> None:
         if self.unit.is_leader():
             logger.info("This app's public ingress URL: %s", event.url)
 
         self._handle_status_update_config(event)
         self._update_hydra_endpoints_relation_data(event)
+        self._update_endpoint_info()
 
     def _on_ingress_revoked(self, event: IngressPerAppRevokedEvent) -> None:
         if self.unit.is_leader():
             logger.info("This app no longer has ingress")
 
         self._handle_status_update_config(event)
         self._update_hydra_endpoints_relation_data(event)
+        self._update_endpoint_info()
+
+    def _on_oauth_relation_created(self, event: RelationCreatedEvent) -> None:
+        self._update_endpoint_info()
+
+    def _on_client_created(self, event: ClientCreatedEvent) -> None:
+        if not self._container.can_connect():
+            event.defer()
+            return
+
+        try:
+            self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
+            event.defer()
+            return
+
+        client_config = event.to_client_config()
+        try:
+            client = self._hydra_cli.create_client(
+                client_config, metadata=f'{{"relation_id": {event.relation_id}}}'
+            )
+        except ExecError as err:
+            logger.error(f"Exited with code: {err.exit_code}. Stderr: {err.stderr}")
+            return
+        except Error as e:
+            logger.error(f"Something went wrong when trying to run the command: {e}")
+            logger.info("Deferring the event")
+            event.defer()
+            return
+
+        self._set_oauth_relation_peer_data(event.relation_id, dict(client_id=client["client_id"]))
+        self.oauth.set_client_credentials_in_relation_data(
+            event.relation_id, client["client_id"], client["client_secret"]
+        )
+
+    def _on_client_changed(self, event: ClientChangedEvent) -> None:
+        if not self._container.can_connect():
+            event.defer()
+            return
+
+        try:
```

Comment:
I see this try/except in a couple of places. You may want to create a `_service_is_running(self, service_name: str)` method and call it here:

```python
if not self._service_is_running(service_name):
    event.defer()
    return
```